vpr/src/pack/pack.cpp

```diff
@@ -156,30 +155,45 @@ bool try_pack(t_packer_opts* packer_opts,
             clustering_data);
 
         //Try to size/find a device
-        bool fits_on_device = try_size_device_grid(*arch, helper_ctx.num_used_type_instances, packer_opts->target_device_utilization, packer_opts->device_layout);
+        std::vector<std::string> overused_blocks = try_size_device_grid(*arch, helper_ctx.num_used_type_instances, packer_opts->target_device_utilization, packer_opts->device_layout);
 
         /* We use this bool to determine the cause for the clustering not being dense enough. If the clustering
          * is not dense enough and there are floorplan constraints, it is presumed that the constraints are the cause
          * of the floorplan not fitting, so attraction groups are turned on for later iterations.
          */
         bool floorplan_not_fitting = (floorplan_regions_overfull || g_vpr_ctx.mutable_floorplanning().constraints.get_num_partitions() > 0);
 
-        if (fits_on_device && !floorplan_regions_overfull) {
+        if (overused_blocks.empty() && !floorplan_regions_overfull) {
             break; //Done
         } else if (pack_iteration == 1 && !floorplan_not_fitting) {
             //1st pack attempt was unsucessful (i.e. not dense enough) and we have control of unrelated clustering
             //
             //Turn it on to increase packing density
-            if (packer_opts->allow_unrelated_clustering == e_unrelated_clustering::AUTO) {
-                VTR_ASSERT(allow_unrelated_clustering == false);
-                allow_unrelated_clustering = true;
+            std::vector<std::string> fixed_blocks;
+            for (auto& overused_block : overused_blocks) {
+                e_unrel_clust_mode block_mode = allow_unrelated_clustering.get_block_mode(overused_block);
+                if (block_mode == e_unrel_clust_mode::FIXED) {
+                    // The user has explicitly set unrelated clustering status for this block type
+                    // so we can't change it
+                    fixed_blocks.push_back(overused_block);
+                } else {
+                    allow_unrelated_clustering.set_block_status(overused_block, std::make_pair(e_unrel_clust_stat::ON, block_mode));
+                }
             }
             if (packer_opts->balance_block_type_utilization == e_balance_block_type_util::AUTO) {
                 VTR_ASSERT(balance_block_type_util == false);
                 balance_block_type_util = true;
             }
-            VTR_LOG("Packing failed to fit on device. Re-packing with: unrelated_logic_clustering=%s balance_block_type_util=%s\n",
-                    (allow_unrelated_clustering ? "true" : "false"),
+
+            if (fixed_blocks.size() > 0) {
+                std::stringstream msg;
+                msg << "Packing failed to fit on device.\n"
+                    << "The following block types were overused, but unrelated clustering was disabled for them:\n"
```

Comment:
Is unrelated clustering set for them, or just a fixed mode (either on or off)?  I think this message isn't quite right.